app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/LocalFileTransferActivity.kt

```diff
@@ -0,0 +1,384 @@
+import kotlinx.android.synthetic.main.activity_local_file_transfer.progress_bar_searching_peers
+import kotlinx.android.synthetic.main.activity_local_file_transfer.recycler_view_transfer_files
+import kotlinx.android.synthetic.main.activity_local_file_transfer.text_view_device_name
+import kotlinx.android.synthetic.main.activity_local_file_transfer.text_view_empty_peer_list
+import org.kiwix.kiwixmobile.R
+import org.kiwix.kiwixmobile.core.base.BaseActivity
+import org.kiwix.kiwixmobile.core.di.components.CoreComponent
+import org.kiwix.kiwixmobile.core.extensions.toast
+import org.kiwix.kiwixmobile.core.utils.dialog.AlertDialogShower
+import org.kiwix.kiwixmobile.core.utils.dialog.KiwixDialog
+import org.kiwix.kiwixmobile.kiwixActivityComponent
+import org.kiwix.kiwixmobile.localFileTransfer.WifiDirectManager.Companion.getDeviceStatus
+import org.kiwix.kiwixmobile.localFileTransfer.adapter.WifiP2pDelegate
+import org.kiwix.kiwixmobile.localFileTransfer.adapter.WifiPeerListAdapter
+import java.util.ArrayList
+import javax.inject.Inject
+
+/**
+ * Created by @Aditya-Sood as a part of GSoC 2019.
+ *
+ * This activity is the starting point for the module used for sharing zims between devices.
+ *
+ * The module is used for transferring ZIM files from one device to another, from within the
+ * app. Two devices are connected to each other using WiFi Direct, followed by file transfer.
+ *
+ * File transfer involves two phases:
+ * 1) Handshake with the selected peer device, using [PeerGroupHandshakeAsyncTask]
+ * 2) After handshake, starting the files transfer using [SenderDeviceAsyncTask] on the sender
+ * device and [ReceiverDeviceAsyncTask] files receiving device
+ */
+@SuppressLint("GoogleAppIndexingApiWarning", "Registered")
+class LocalFileTransferActivity : BaseActivity(),
+  WifiDirectManager.Callbacks {
+  @Inject
+  lateinit var alertDialogShower: AlertDialogShower
+
+  @Inject
+  lateinit var wifiDirectManager: WifiDirectManager
+
+  @Inject
+  lateinit var locationManager: LocationManager
+
+  private var fileListAdapter: FileListAdapter? = null
+  private var wifiPeerListAdapter: WifiPeerListAdapter? = null
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_local_file_transfer)
+    /*
+     * Presence of file Uris decides whether the device with the activity open is a sender or receiver:
+     * - On the sender device, this activity is started from the app chooser post selection
+     * of files to share in the Library
+     * - On the receiver device, the activity is started directly from within the 'Get Content'
+     * activity, without any file Uris
+     * */
+    val filesIntent = intent
+    val fileUriArrayList: ArrayList<Uri>? =
+      filesIntent.getParcelableArrayListExtra(Intent.EXTRA_STREAM)
+    val isFileSender = fileUriArrayList?.isNotEmpty() == true
+    val filesForTransfer: List<FileItem>? = if (isFileSender)
```

Comment:
lets change this up a tiny bit to avoid this nullability lark.

```
val fileForTransfer = fileUriArrayList?.map(::FileItem) ?: emptyList()
val isFileSender = filesForTransfer.isNotEmpty()
```

now you won't need the `?.let`s.

Do we even need `isFileSender` at all?

Reply:
yes, I had a similar hunch about `isFileSender` .